--- node_modules/node-xml/lib/node-xml.js	2011-08-23 09:01:59.013045629 +0200
+++ node_modules/node-xml/lib/node-xml_fix_rain.js	2011-08-23 09:02:56.243047653 +0200
@@ -665,6 +665,7 @@
     iB = iB || 0;
     iE = iE || strD.length;
 
+    /*
     switch(strD.substring(iB, iE)) {
         case "amp":  strEnt = "&";  break;
         case "lt":   strEnt = "<";  break;
@@ -691,6 +692,9 @@
             }
         break;
     }
+    */
+    strEnt = '&' + strD.substring(iB, iE) + ';'
+
     this._setContent(XMLP._CONT_ALT, strEnt);
 
     return XMLP._ENTITY;
