// socket.io/lib-cov/transports/http-polling.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['transports/http-polling.js']) {
  _$jscoverage['transports/http-polling.js'] = [];
  _$jscoverage['transports/http-polling.js'][12] = 0;
  _$jscoverage['transports/http-polling.js'][18] = 0;
  _$jscoverage['transports/http-polling.js'][26] = 0;
  _$jscoverage['transports/http-polling.js'][27] = 0;
  _$jscoverage['transports/http-polling.js'][28] = 0;
  _$jscoverage['transports/http-polling.js'][36] = 0;
  _$jscoverage['transports/http-polling.js'][42] = 0;
  _$jscoverage['transports/http-polling.js'][43] = 0;
  _$jscoverage['transports/http-polling.js'][52] = 0;
  _$jscoverage['transports/http-polling.js'][53] = 0;
  _$jscoverage['transports/http-polling.js'][55] = 0;
  _$jscoverage['transports/http-polling.js'][56] = 0;
  _$jscoverage['transports/http-polling.js'][58] = 0;
  _$jscoverage['transports/http-polling.js'][59] = 0;
  _$jscoverage['transports/http-polling.js'][60] = 0;
  _$jscoverage['transports/http-polling.js'][63] = 0;
  _$jscoverage['transports/http-polling.js'][73] = 0;
  _$jscoverage['transports/http-polling.js'][74] = 0;
  _$jscoverage['transports/http-polling.js'][75] = 0;
  _$jscoverage['transports/http-polling.js'][76] = 0;
  _$jscoverage['transports/http-polling.js'][77] = 0;
  _$jscoverage['transports/http-polling.js'][80] = 0;
  _$jscoverage['transports/http-polling.js'][89] = 0;
  _$jscoverage['transports/http-polling.js'][90] = 0;
  _$jscoverage['transports/http-polling.js'][92] = 0;
  _$jscoverage['transports/http-polling.js'][101] = 0;
  _$jscoverage['transports/http-polling.js'][102] = 0;
  _$jscoverage['transports/http-polling.js'][111] = 0;
  _$jscoverage['transports/http-polling.js'][112] = 0;
  _$jscoverage['transports/http-polling.js'][113] = 0;
  _$jscoverage['transports/http-polling.js'][114] = 0;
  _$jscoverage['transports/http-polling.js'][123] = 0;
  _$jscoverage['transports/http-polling.js'][124] = 0;
  _$jscoverage['transports/http-polling.js'][125] = 0;
}
_$jscoverage['transports/http-polling.js'][12]++;
var HTTPTransport = require("./http");
_$jscoverage['transports/http-polling.js'][18]++;
exports = module.exports = HTTPPolling;
_$jscoverage['transports/http-polling.js'][26]++;
function HTTPPolling(mng, data) {
  _$jscoverage['transports/http-polling.js'][27]++;
  HTTPTransport.call(this, mng, data);
}
_$jscoverage['transports/http-polling.js'][28]++;
;
_$jscoverage['transports/http-polling.js'][36]++;
HTTPPolling.prototype.__proto__ = HTTPTransport.prototype;
_$jscoverage['transports/http-polling.js'][42]++;
HTTPPolling.prototype.setHeartbeatInterval = (function () {
  _$jscoverage['transports/http-polling.js'][43]++;
  return this;
});
_$jscoverage['transports/http-polling.js'][52]++;
HTTPPolling.prototype.handleRequest = (function (req) {
  _$jscoverage['transports/http-polling.js'][53]++;
  HTTPTransport.prototype.handleRequest.call(this, req);
  _$jscoverage['transports/http-polling.js'][55]++;
  if (req.method == "GET") {
    _$jscoverage['transports/http-polling.js'][56]++;
    var self = this;
    _$jscoverage['transports/http-polling.js'][58]++;
    this.pollTimeout = setTimeout((function () {
  _$jscoverage['transports/http-polling.js'][59]++;
  self.close();
  _$jscoverage['transports/http-polling.js'][60]++;
  self.log.debug("polling closed due to exceeded duration");
}), this.manager.get("polling duration") * 1000);
    _$jscoverage['transports/http-polling.js'][63]++;
    this.log.debug("setting poll timeout");
  }
});
_$jscoverage['transports/http-polling.js'][73]++;
HTTPPolling.prototype.clearPollTimeout = (function () {
  _$jscoverage['transports/http-polling.js'][74]++;
  if (this.pollTimeout) {
    _$jscoverage['transports/http-polling.js'][75]++;
    clearTimeout(this.pollTimeout);
    _$jscoverage['transports/http-polling.js'][76]++;
    this.pollTimeout = null;
    _$jscoverage['transports/http-polling.js'][77]++;
    this.log.debug("clearing poll timeout");
  }
  _$jscoverage['transports/http-polling.js'][80]++;
  return this;
});
_$jscoverage['transports/http-polling.js'][89]++;
HTTPPolling.prototype.clearTimeouts = (function () {
  _$jscoverage['transports/http-polling.js'][90]++;
  HTTPTransport.prototype.clearTimeouts.call(this);
  _$jscoverage['transports/http-polling.js'][92]++;
  this.clearPollTimeout();
});
_$jscoverage['transports/http-polling.js'][101]++;
HTTPPolling.prototype.doWrite = (function () {
  _$jscoverage['transports/http-polling.js'][102]++;
  this.clearPollTimeout();
});
_$jscoverage['transports/http-polling.js'][111]++;
HTTPPolling.prototype.write = (function (data, close) {
  _$jscoverage['transports/http-polling.js'][112]++;
  this.doWrite(data);
  _$jscoverage['transports/http-polling.js'][113]++;
  this.response.end();
  _$jscoverage['transports/http-polling.js'][114]++;
  this.onClose();
});
_$jscoverage['transports/http-polling.js'][123]++;
HTTPPolling.prototype.end = (function () {
  _$jscoverage['transports/http-polling.js'][124]++;
  this.clearPollTimeout();
  _$jscoverage['transports/http-polling.js'][125]++;
  return HTTPTransport.prototype.end.call(this);
});
_$jscoverage['transports/http-polling.js'].source = ["","/*!"," * socket.io-node"," * Copyright(c) 2011 LearnBoost &lt;dev@learnboost.com&gt;"," * MIT Licensed"," */","","/**"," * Module requirements."," */","","var HTTPTransport = require('./http');","","/**"," * Exports the constructor."," */","","exports = module.exports = HTTPPolling;","","/**"," * HTTP polling constructor."," *"," * @api public."," */","","function HTTPPolling (mng, data) {","  HTTPTransport.call(this, mng, data);","};","","/**"," * Inherits from HTTPTransport."," *"," * @api public."," */","","HTTPPolling.prototype.__proto__ = HTTPTransport.prototype;","","/**"," * Removes heartbeat timeouts for polling."," */","","HTTPPolling.prototype.setHeartbeatInterval = function () {","  return this;","};","","/**"," * Handles a request"," *"," * @api private"," */","","HTTPPolling.prototype.handleRequest = function (req) {","  HTTPTransport.prototype.handleRequest.call(this, req);","","  if (req.method == 'GET') {","    var self = this;","","    this.pollTimeout = setTimeout(function () {","      self.close();","      self.log.debug('polling closed due to exceeded duration');","    }, this.manager.get('polling duration') * 1000);","","    this.log.debug('setting poll timeout');","  }","};","","/**"," * Clears polling timeout"," *"," * @api private"," */","","HTTPPolling.prototype.clearPollTimeout = function () {","  if (this.pollTimeout) {","    clearTimeout(this.pollTimeout);","    this.pollTimeout = null;","    this.log.debug('clearing poll timeout');","  }","","  return this;","};","","/**"," * Override clear timeouts to clear the poll timeout"," *"," * @api private"," */","","HTTPPolling.prototype.clearTimeouts = function () {","  HTTPTransport.prototype.clearTimeouts.call(this);","","  this.clearPollTimeout();","};","","/**"," * doWrite to clear poll timeout"," *"," * @api private"," */","","HTTPPolling.prototype.doWrite = function () {","  this.clearPollTimeout();","};","","/**"," * Performs a write."," *"," * @api private."," */","","HTTPPolling.prototype.write = function (data, close) {","  this.doWrite(data);","  this.response.end();","  this.onClose();","};","","/**"," * Override end."," *"," * @api private"," */","","HTTPPolling.prototype.end = function () {","  this.clearPollTimeout();","  return HTTPTransport.prototype.end.call(this);","};",""];
